### ISSUE TITLE
verify chosen address is registered

### ISSUE THREAD

**Issue by dckc** [rchain-community/rv2020#77]
earlier we noted that the list should not go out until voting is closed, else people can do tallying while voting is open
https://github.com/rchain-community/rv2020/issues/35#issuecomment-690469525

@jimscarver points out a risk that a registered voter will mistakenly pick the wrong address and that the voting tool could prevent this if the list of registered REV addresses went on chain earlier.

**Comment by dckc** [comment]
There's a policy question here. @jimscarver @SteveHenley @ian-bloom what do you think? Or would you prefer to consult with the rest of the gov cttee before answering?

**Comment by jimscarver** [comment]
my suggestion is that the member be told up front if the address is unregistered. I do not consider the voter list to be private since it will be revealed after the vote is over anyway. Not revealing it before could be interpreted as nefarious.

**Comment by jimscarver** [comment]
people will tally in any case. Why is tallying before the election close bad? It is no better than a voter survey.

**Comment by dckc** [comment]
@jimscarver I'm aware of your preference. But it runs counter to what I understood was the consensus from [our Sep 10 rv2020/rchat discussion](https://github.com/rchain-community/rv2020/issues/35#issuecomment-690469525), so please get input from either that group or the gov cttee.

**Comment by jimscarver** [comment]
I will check with ian on when he plans to send list of registered addresses. If there is any issue governance will address next week

**Comment by TheoXD** [comment]
What if co-op releases hashed versions of member addresses, would that help? And another issue is that we can assume that most transactions coming in to those addresses will be from members, so you would be able to see results early. A solution to that would be to generate some noise by sending rev from non-member wallets.

**Comment by dckc** [comment]
> What if co-op releases hashed versions of member addresses, would that help?

No, on further thought, I remember why not. Suppose the secretary does:

```python
>>> registeredAddrs = [...]
>>> hashedAddrs = [hash(va) for va in registeredAddrs]
>>> regURI = rchain.register(hashedAddrs)
```

Then the "attacker" can do:

```python
>>> dests = [... from the ballot ...]
>>> maybeVotes = [tx for tx in indexService() if tx.recipient in dests]
>>> hashedAddrs = lookup("rho:...")
>>> realVotes = [tx for tx in maybeVotes if hash(tx.sender) in hashedAddrs]
```

IOW: given a set `reg` of `h(va)` for each registered voter address `va`, the "attacker" who wants to tally the votes early can find all the votes given the ballot and then look at the sender address `sa` of each vote, compute `h(sa)` and check whether it's in `reg`.

If anyone still thinks this is worth pursuing, would you please do an experiment where you
  - make up a few REV addresses
  - hash them
  - put them on chain with `insertArbitrary`
  - send me the `rho:` URI
?

Also, consult with @lrusu and @ian-bloom on what sort of tool they would like to use. Is a nodejs thing cool? Or a bash script? Sometimes a google spreadsheet is a great UI, but maybe we don't want to share the list with Google?

**Comment by ian-bloom** [comment]
Currently, an "attacker" can monitor the progress of the election by watching yes/no ballot addresses as well as sender addresses. Sender addresses are exposed 3 ways:

1. Co-op funds voter's addresses with dust prior to the election
2. receipt of REV by yes/no ballot question addresses
3. REV addresses that received Co-op's 100 REV early registration bonus

As @TheoXD suggests, we could obscure the vote by (1) funding fake voter addresses from which dust is sent to (2), but would still have to ask early registrants for help obscuring (3). This may over-complicate voter instructions.

If obscuring the vote progress is too hard to implement in this version, @jimscarver suggestion of publishing the valid sender addresses might be the best way to go. It seems unlikely that we could implement encrypted voting or contract voting in time for this year's election.

**Comment by dckc** [comment]
We (Sep 26 attendees) are RESOLVED that the benefit to publishing the voter rev addresses at the start of the voting period outweighs the costs / risks.

"... no later than oct 20"

**Comment by dckc** [comment]
@jimscarver in https://github.com/rchain-community/rv2020/issues/34#issuecomment-699661139 
you refer to `rho:id:kiijxigqydnt7ds3w6w3ijdszswfysr3hpspthuyxz4yn3ksn4ckzf`

I don't find anything under that URI in the testnet registry. Is that from mainnet?

**Comment by jimscarver** [comment]
I added abstain on the ballot-w-abstain branch at the addresses below. I return both the ballot and if they are registered or not but am unable to debug because metamask is no longer finding the rnode-client-js stuff. I tried a symbolic link in the vendor directory but that does not help.
```scala
new return, lookup(`rho:registry:lookup`)
in {
  new valueCh in {
    lookup!( `rho:id:5rcmyxwu8r7yywjz4qqg4ij3pox3d96joeky1gczdpf3fkaujejdxr` , *valueCh) |
    for (@ballot <- valueCh) {
        lookup!( `rho:id:1ri71weozwuoanef9zit5p7ooafkmkzhkwo6phgaourcknbmi6ke7t` , *valueCh) |
        for (@accts <- valueCh) {
          match accts {
                  { "11111LHTzJGYkvmr8zK4FwNTy1Bbrmu5QnS91L13Azb6xNEu7a6gb" | rest } => {
                      return!({"registered": true ,"ballot": ballot})
                  }
                  { _ } => {
                      return!({"registered": false ,"ballot": ballot})
                  }
              }
          }
      }
  }
}
```
Result
```js
[
  {
    "ballot": {
      "Board: WEC": {
        "yesAddr": "11112gUFvJR6JBDYJURETaWUBpEDa1EyjgRHFncEfQ4hGECnciPnhw",
        "noAddr": "11112aoa6NLYomYZro566XZVGEXyCDqeqDcp8Pzg81Ckuws6SexC99",
        "shortDesc": "Wile E. Coyote for Board Member",
        "docLink": "https://gist.github.com/dckc/ca240e5336d0ee3e4f5cf31c4f629a30#board-wec",
        "abstainAddr": "1111pKehMgsPBAiqzCSkSekXP4aUXMjY5DvtSXcz72ATP7Pm3RK9o"
      },
      "Board: DaD": {
        "yesAddr": "1111TnFUN7eZBWXp3QQACQRRxpcS5uH5Bpf67vikWhA5e3F6ikAmU",
        "noAddr": "11112Cwtg2Bs4WUAYrXhL9xZXXSXr9Gn62Cty39RhUaBnqjrKkqwAZ",
        "shortDesc": "Daffy Duck for Board Member",
        "docLink": "https://gist.github.com/dckc/ca240e5336d0ee3e4f5cf31c4f629a30#board-dad",
        "abstainAddr": "11112nT2XooHcCVQLEAsEJhQm6boCS5B7XQ1DBmw6ex3xveiCWRWAx"
      },
      "Member Swag": {
        "yesAddr": "11112i8bYVDYcm4MSbY3d1As28uY151xoMS7AyiTvZ2YmNJ8Nw13v9",
        "noAddr": "11112uGayGEi57D44Drq3V4iw5WWyfXbcVvsDangRTE7TaR3J4U4FD",
        "shortDesc": "The Item of Business I want to propose is to provide all new members with stickers and t-shirts with the RChain logo on it as part of their membership onboarding package.",
        "docLink": "https://gist.github.com/dckc/ca240e5336d0ee3e4f5cf31c4f629a30#member-swag",
        "abstainAddr": "111184Ab7raMAoVy6fX8JuoPFB5PggfrEWfzXE4WMzTKioFwmQMsa"
      },
      "Board: DoD": {
        "yesAddr": "1111rbdV9Lsw6DyMSq8ySXDacX7pRUxmVGoYho9gGtfZcQYFdAN42",
        "noAddr": "1111JoeZHDYXqyAgo89VaidQnp7W7M9pvdkFUJTqEBU7SHKx6WF2z",
        "shortDesc": "Donald Duck for Board Member",
        "docLink": "https://gist.github.com/dckc/ca240e5336d0ee3e4f5cf31c4f629a30#board-dod",
        "abstainAddr": "11113Y89LxqCmjDK9PUDi1dfsEcjAHbBW7mQ3Zw2yqqiwSUibaTkq"
      },
      "Board: RR": {
        "yesAddr": "1111krbAKSbyGA9vfa7w4K2pKAxZZn6qjaVEduDLWotDZ8HLt2aXR",
        "noAddr": "1111swBFUPVRwR4ugkDBCvrLwPeR1621B1cHQf3cAkNxt3Zad2eac",
        "shortDesc": "Road Runner for Board Member",
        "docLink": "https://gist.github.com/dckc/ca240e5336d0ee3e4f5cf31c4f629a30#board-rr",
        "abstainAddr": "11112CgGiNg3DdMDsYz7UikeSxh7CfFdEbDYzmoJLfS4vx3uZjm55V"
      }
    },
    "registered": true
  }
]
```
getting the ballot and flag out of the result and populating the contract with the abstainAddr default still need to be done. I am hoping the debugger can help me finish if I can get it running again.

**Comment by jimscarver** [comment]
The problem with storing sets was my error so I am reverting to use the Set of rev addresses stored on the chain.
```scala
new return, lookup(`rho:registry:lookup`)
in {
  new valueCh in {
    // Fill in registry URI: `rho:id:11fhnau8j3...h4459w9bpus6oi`
    lookup!( `rho:id:kiijxigqydnt7ds3w6w3ijdszswfysr3hpspthuyxz4yn3ksn4ckzf` , *valueCh) |
    for (@value <- valueCh) {
      return!((value.contains("1111tBR2JPWkpRmXXqsRU4aCK3rMcoBFgjEQpTJKED6VxYt7WuLi7")))
    }
  }
}
```

**Comment by dckc** [comment]
fixed in 50cf6ca56ec13eba4bdc3052e1603e369fa3abae
